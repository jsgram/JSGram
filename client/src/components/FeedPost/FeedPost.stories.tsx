import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { MemoryRouter } from 'react-router-dom';
import { FeedPost } from './index';
import { Provider } from 'react-redux';
import { store } from '../../App';

export const feedPostState = {
    loggedId: '1a',
    loggedUsername: 'Maks',
    loggedPhotoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
    feedLoaded: false,
    feedLoading: false,
    newsFeed: {
        feedLoaded: true,
        feedLoading: false,
        page: 1,
        feed: [
            {
                _id: '1p',
                author: {
                    _id: '1a',
                    username: 'Maks',
                    photoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
                },
                authorsOfLike: ['1f', '2f', '3f'],
                comments: [],
                description: 'Post description',
                imgPath: 'https://jsgram-post-images.s3.eu-west-2.amazonaws.com/1570605728486',
                tags: [],
            },
            {
                _id: '2p',
                author: {
                    _id: '1a',
                    username: 'Maks',
                    photoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
                },
                authorsOfLike: [],
                comments: [],
                description: 'Post description',
                imgPath: 'https://jsgram-post-images.s3.eu-west-2.amazonaws.com/1570605706200',
                tags: [],
            },
            {
                _id: '3p',
                author: {
                    _id: '1a',
                    username: 'Maks',
                    photoPath: 'https://jsgram-profile-images.s3.eu-west-3.amazonaws.com/1569138592191',
                },
                authorsOfLike: [],
                comments: [],
                description: 'Post description',
                imgPath: 'https://jsgram-post-images.s3.eu-west-2.amazonaws.com/1570603193496',
                tags: [],
            },
        ],
        friendsRecommendations: {
            loading: false,
            users: [],
            page: 1,
        },
    },
    friendsRecommendations: {
        loading: false,
        users: [],
    },
};

export const feedPostActions = {
    getNewsFeedAsync: action('getNewsFeedAsync'),
    getMoreNewsFeedAsync: action('getMoreNewsFeedAsync'),
    addNextFeedPosts: action('addNextFeedPosts'),
    getRecommendations: action('getRecommendations'),
    followUser: action('followUser'),
    changeUsersFollowing: action('changeUsersFollowing'),
};

storiesOf('Feed Post', module)
    .addDecorator((story: any): any => <Provider store={store}>{story()}</Provider>)
    .addDecorator((story: any): any => <MemoryRouter>{story()}</MemoryRouter>)
    .add('feed post', () => <FeedPost {...{...feedPostState, ...feedPostActions}}/>);
