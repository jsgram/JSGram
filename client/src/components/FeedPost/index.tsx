import React from 'react';
import '../../styles/style.scss';
import '../Post/style.scss';
import noAvatar from '../../assets/noAvatar.png';
import { Link } from 'react-router-dom';
import { formatDescription } from '../../helpers/regex.description';
import FeedLikesContainer from '../../containers/FeedLikesContainer';
import Menu from '../Menu';
import { Col, Container, Row, Spinner } from 'reactstrap';
import { INewsFeed } from '../../store/newsFeed/reducers';
import { Waypoint } from 'react-waypoint';
import Comment from '../Comments';
import { FriendsRecomendations } from '../FriendsRecommendations';
import WriteComment from '../WriteComment';

interface IProps {
    loggedId: string;
    loggedUsername: string;
    loggedPhotoPath: string;
    newsFeed: any;
    getNewsFeedAsync: () => void;
    getMoreNewsFeedAsync: (page: number) => void;
    addNextFeedPosts: (pageNumber: number) => void;
    getRecommendations: () => void;
    followUser: (body: { _id: string }) => void;
    friendsRecommendations: any;
}

export class FeedPost extends React.Component<IProps> {
    public componentDidMount(): void {
        this.props.getNewsFeedAsync();
        this.props.getRecommendations();
    }

    public getMoreFeedPosts = (): void => {
        if (!this.props.newsFeed.feedLoaded) {
            this.props.addNextFeedPosts(this.props.newsFeed.page);
            this.props.getMoreNewsFeedAsync(this.props.newsFeed.page);
        }
    }

    public render(): JSX.Element {
        const {loggedUsername, loggedPhotoPath, newsFeed, followUser}: any = this.props;

        return (
            <Container>
                <Menu/>
                <Row>
                    <Col sm={12} md={8} className='order-2 order-md-1'>
                        {
                            newsFeed.feed.filter((feed: INewsFeed) => feed._id).map((feed: INewsFeed) => {
                                const {description, imgPath, author: {_id: authorId, photoPath, username}}: any = feed;
                                return (
                                    <div className='profile-post border mb-5' key={feed._id}>
                                        <div className='post-header p-2 border'>
                                            <div className='d-flex flex-row'>
                                                <img
                                                    src={photoPath || noAvatar}
                                                    alt='avatar'
                                                    width={32}
                                                    height={32}
                                                    className='img-fluid rounded-circle'
                                                />
                                                <Link
                                                    to={`/profile/${username}`}
                                                    className='text-dark mt-1 ml-3'>
                                                    {username}
                                                </Link>
                                            </div>
                                        </div>
                                        <img
                                            src={imgPath}
                                            className='w-100 img-fluid'
                                            alt='post'
                                        />
                                        <div className='d-block mt-3 mb-2 pl-3'>
                                            <FeedLikesContainer
                                                postId={feed._id}
                                                likes={feed.authorsOfLike}
                                                authorId={authorId}
                                            />
                                        </div>
                                        <div className='description-post pb-3'>
                                            <div className='d-block pl-3 text-description'>
                                                <img
                                                    src={photoPath || noAvatar}
                                                    alt='avatar'
                                                    width={32}
                                                    height={32}
                                                    className='img-fluid rounded-circle'
                                                />
                                                <Link
                                                    to={`/profile/${username}`}
                                                    className='d-inline-block text-dark ml-2'
                                                >
                                                    {username}
                                                </Link>
                                                <p className='pl-2 mt-2 justify-self-start align-self-start'>
                                                    {formatDescription(description)}
                                                </p>
                                            </div>
                                        </div>
                                        <Comment postId={feed._id}/>
                                            <WriteComment
                                                postId={feed._id}
                                                authorId={feed.author}
                                            />
                                    </div>
                                );
                            })
                        }
                    </Col>
                    <Col sm={12} md={4} className='order-1 order-md-2 text-center'>
                        <div className='sticky-top'>
                            <img
                                src={loggedPhotoPath || noAvatar}
                                alt='avatar'
                                width={64}
                                height={64}
                                className='img-fluid rounded-circle'
                            />
                            <Link to={`/profile/${loggedUsername}`} className='mt-1 ml-3 mr-4
                        text-dark'>{loggedUsername}</Link>
                            <Link to='/logout' className='text-danger pl-1'>Logout</Link>
                            {!!newsFeed.friendsRecommendations.users.length &&
                            <FriendsRecomendations
                                loggedUsername={loggedUsername}
                                friendsRecommendations={newsFeed.friendsRecommendations}
                                followUser={followUser}
                            />}
                        </div>
                    </Col>
                </Row>

                <Waypoint
                    scrollableAncestor={window}
                    onEnter={(): void => {
                        this.getMoreFeedPosts();
                    }}
                />
                <div className='w-100 d-flex align-items-center justify-content-center'>
                    {this.props.newsFeed.feedLoading && <Spinner className='mt-3' color='dark'/>}
                </div>
            </Container>
        );
    }
}
