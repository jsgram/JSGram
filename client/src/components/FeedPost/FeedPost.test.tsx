import React from 'react';
import { FeedPost } from './index';
import { shallow } from 'enzyme';
import { feedPostState } from './FeedPost.stories';

describe('FeedPost smart component', () => {
    let renderer: any;
    const actions = {
        getNewsFeedAsync: jest.fn(() => 'some value'),
        getMoreNewsFeedAsync: jest.fn(() => 'some value'),
        addNextFeedPosts: jest.fn(() => 'some value'),
        getRecommendations: jest.fn(() => 'some value'),
        followUser: jest.fn(() => 'some value'),
    };

    beforeEach(() => {
        renderer = shallow(<FeedPost {...{...feedPostState, ...actions}} />);
    });

    test('componentDidMount - success', () => {
        renderer.instance().componentDidMount();
        expect(actions.getNewsFeedAsync).toHaveReturnedWith('some value');
        expect(actions.getRecommendations).toHaveReturnedWith('some value');
    });

    test('get more feed posts - success', () => {
        renderer.instance().getMoreFeedPosts();
        expect(actions.addNextFeedPosts).not.toHaveBeenCalled();
        expect(actions.getMoreNewsFeedAsync).not.toHaveBeenCalled();
    });

    test('render - success', () => {
        expect(renderer).toMatchSnapshot();
    });
});
