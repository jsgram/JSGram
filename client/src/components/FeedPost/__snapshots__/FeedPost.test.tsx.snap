// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedPost smart component render - success 1`] = `ShallowWrapper {}`;
